gibbon-compiler/src/Gibbon/Passes/HoistExpressions.hs

```diff
@@ -0,0 +1,340 @@
+module Gibbon.Passes.HoistExpressions (hoistBoundsCheckProg) where
+
+import Data.Foldable (foldlM)
+import qualified Data.Map as M
+import qualified Data.Set as S
+import Gibbon.Common
+import Gibbon.NewL2.Syntax as NewL2
+
+---------------------------------------------------------------------------
+-- Hoist the bounds check expression to the top of each function
+
+-- | Map to store variables in scope
+type BoundEnv = S.Set FreeVarsTy
+
+-- | The type of expressions that can be hoisted to the top
+data HoistableExpr
+  = BoundsCheckExpr Int LocArg LocArg
+  | BoundsCheckVectorExpr [(Int, LocArg, LocArg)]
+  | LetLocExpr LocVar (PreLocExp LocArg)
+  | LetRegExpr RegVar (PreRegExp LocArg)
+  | LetExpr (Var, [LocArg], Ty2, PreExp E2Ext LocArg Ty2)
+  | LetRegionExpr Region RegionSize (Maybe RegionType)
+  deriving (Eq, Ord)
+
+-- | Stores all the expressions that can be hoisted to the top of the function
+type HoistAbleExprMap = M.Map HoistableExpr (S.Set FreeVarsTy)
+
+mergeHoistExprMaps :: [HoistAbleExprMap] -> HoistAbleExprMap
+mergeHoistExprMaps = foldr (M.unionWith S.union) M.empty
+
+fromLocArgToFreeVarsTy' :: LocArg -> [FreeVarsTy]
+fromLocArgToFreeVarsTy' arg =
+  case arg of
+    Loc lrm -> [fromLocVarToFreeVarsTy $ lremLoc lrm, fromRegVarToFreeVarsTy $ lremReg lrm]
+    EndWitness _ v -> [fromLocVarToFreeVarsTy v]
+    Reg v _ -> [fromRegVarToFreeVarsTy v]
+    EndOfReg v1 _ v2 -> [fromRegVarToFreeVarsTy v1, fromRegVarToFreeVarsTy v2]
+    EndOfReg_Tagged v -> [fromRegVarToFreeVarsTy v]
+
+collectBoundsCheckExprs :: HoistAbleExprMap -> BoundEnv -> NewL2.Exp2 -> PassM (NewL2.Exp2, HoistAbleExprMap)
```

Comment:
The function name tells that the function collect certain expressions. That would explain the map in the result. But what is up with the Exp2 there? This function, just as every non-trivial function needs a comment.